### PR TITLE
Use unicode-xid crate instead of libcore

### PR DESCRIPTION
**Pull request by matklad**
This PR proposes to remove `char::is_xid_start` and `char::is_xid_continue` functions from `libcore` and use `unicode_xid` crate from crates.io (note that this crate is already present in rust-lang/rust's Cargo.lock). 

Reasons to do this:

* removing rustc-binary-specific stuff from libcore
* making sure that, across the ecosystem, there's a single definition of what rust identifier is (`unicode-xid` has almost 10 million downs, as a `proc_macro2` dependency)
* making it easier to share `rustc_lexer` crate with rust-analyzer: no need to `#[cfg]` if we are building as a part of the compiler

Reasons not to do this:

* increased maintenance burden: we'll need to upgrade unicode version both in libcore and in unicode-xid. However, this shouldn't be a too heavy burden: just running `./unicode.py` after new unicode version. I (@matklad) am ready to be a t-compiler side maintainer of unicode-xid. Moreover, given that xid-unicode is an important dependency of syn, *someone* needs to maintain it anyway.
* xid-unicode implementation is significantly slower. It uses a more compact table with binary search, instead of a trie. However, this shouldn't matter in practice, because we have fast-path for ascii anyway, and code size savings is a plus. Moreover, in #59706 not using libcore turned out to be *faster*, presumably beacause checking for whitespace with match is even faster. 

<details>

<summary>old description</summary>

Followup to #59706 

r? @eddyb

Note that this doesn't actually remove tables from libcore, to avoid conflict with https://github.com/rust-lang/rust/pull/62641.

cc https://github.com/unicode-rs/unicode-xid/pull/11

</details>

### CODE REVIEW

File: src/libfmt_macros/lib.rs [hunk 1]
```diff
@@ -575,12 +576,11 @@ impl<'a> Parser<'a> {
         }
     }
 
-    /// Parses a word starting at the current position. A word is considered to
-    /// be an alphabetic character followed by any number of alphanumeric
-    /// characters.
+    /// Parses a word starting at the current position. A word is the same as
+    /// Rust identifier, except that it can't start with `_` character.
     fn word(&mut self) -> &'a str {
```
Comment (matklad) [inline src/libfmt_macros/lib.rs]: Note that we (accidentally?) don't gate on non-ascii idents here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c5673f4b78f88393d9ba4f46f4838d38

**Comment by petrochenkov** [comment]
>Note that this doesn't actually remove tables from libcore, to avoid conflict with #62641.

Well, one rebase seems better than a separate PR.

**Comment by matklad** [comment]
Makes sense, added a commit that regenerates table using existing unicode version. 

However, removing `char::is_xid_start` from libcore broke the build spectacularly: it was the only place that declared `#[feature = "rustc_private"]`, and compiler now complains that it doesn't know of such a feature, although it is used everywhere.

**Comment by matklad** [comment]
cc https://github.com/rust-lang/rust/pull/62869

**Comment by matklad** [comment]
So the test now fail because running `./unicode.py --version 11.0.0` somehow changed `Grapheme_Extend_table` ... Not sure what's up with this, the simplest thing to do would be to wait until #62641 

@rustbot modify labels to +S-blocked -S-waiting-on-author

**Comment by matklad** [comment]
This is now blocked on https://github.com/unicode-rs/unicode-xid/pull/11.

cc @Manishearth :-)

**Comment by matklad** [comment]
@rfcbot fcp merge

**Comment by Mark-Simulacrum** [comment]
@rfcbot ask @rust-lang/compiler Should we move to using unicode-xid?

**Comment by estebank** [comment]
Let's have a crater run just to verify that the change to use `is_id_start` is not user visible.

**Comment by varkor** [comment]
Just to clarify: is `xid-unicode` going to be moved under https://github.com/rust-lang, or will it remain a 3rd-party crate?

**Comment by matklad** [comment]
@varkor I assume that it remains a third party crate, but moving it to rust-lang/rust is also an option I think, let’s ask @Manishearth :)

**Comment by Manishearth** [comment]
Probably stay third-party. Depends on what unicode-rs wants.

**Comment by matklad** [comment]
So looks like [there's consensus](https://github.com/rust-lang/rust/pull/62848#issuecomment-515497395) that we should move forward with this. 

I've rebased the changes, should be ready for review.

r? @petrochenkov 

@rustbot modify labels: -S-blocked, +S-waiting-on-review

File: src/librustc_lexer/src/lib.rs [hunk 2]
```diff
@@ -103,6 +102,62 @@ pub fn tokenize(mut input: &str) -> impl Iterator<Item = Token> + '_ {
...
+/// True if `c` is valid as a non-first character of an identifier.
+pub fn is_id_continue(c: char) -> bool {
+    // This is exactly XID_Continue.
+    // We also add fast-path for ascii idents
+    ('a' <= c && c <= 'z')
+        || ('A' <= c && c <= 'Z')
+        || ('0' <= c && c <= '9')
+        || c == '_'
+        || (c > '\x7f' && unicode_xid::UnicodeXID::is_xid_continue(c))
+}
```
Comment (matklad) [inline src/librustc_lexer/src/lib.rs]: Should I mark these three functions as `#[inline]`? I imaging being able to inline ascii fast-path might be important for perf, if these functions are hot. 

On the other hand, I don't expect that they are hot anywhere outside the lexer, and the lexer's code is in the same crate so it already can do inlining. For this reason, I will *not* add `#[inline]`, unless somone thinks that we should do it.

Comment (petrochenkov) [reply src/librustc_lexer/src/lib.rs]: >I don't expect that they are hot anywhere outside the lexer, and the lexer's code is in the same crate so it already can do inlining.

Agreed.

**Pull request merged**